File: python/tests/test_highscore.py
import json
import os
import tempfile
import unittest

from snake_game.highscore import HighScoreStore


class HighScoreStoreTests(unittest.TestCase):
    def setUp(self):
        fd, self.path = tempfile.mkstemp(suffix='.json')
        os.close(fd)
        os.remove(self.path)  # start with no save file

    def tearDown(self):
        if os.path.exists(self.path):
            os.remove(self.path)

    def test_defaults_to_zero_when_no_file_exists(self):
        store = HighScoreStore(self.path)
        self.assertEqual(store.value, 0)

    def test_update_persists_a_new_high_score_to_disk(self):
        store = HighScoreStore(self.path)
        store.update(12)
        with open(self.path) as f:
            self.assertEqual(json.load(f)['high_score'], 12)

    def test_update_ignores_a_lower_score(self):
        store = HighScoreStore(self.path)
        store.update(12)
        store.update(5)
        self.assertEqual(store.value, 12)

    def test_a_new_store_reloads_the_persisted_value(self):
        store = HighScoreStore(self.path)
        store.update(9)
        reloaded = HighScoreStore(self.path)
        self.assertEqual(reloaded.value, 9)


if __name__ == '__main__':
    unittest.main()

File: cpp/tests/test_highscore.cpp
#include <cstdio>
#include <fstream>
#include <string>

#include "doctest.h"
#include "highscore.hpp"

namespace {

// unique scratch path per test run
std::string TempPath() {
    static int counter = 0;
    return "test_highscore_" + std::to_string(counter++) + ".json";
}

}  // namespace

TEST_CASE("defaults to zero when no file exists") {
    std::string path = TempPath();
    std::remove(path.c_str());
    HighScoreStore store(path);
    CHECK(store.Value() == 0);
}

TEST_CASE("update persists a new high score to disk") {
    std::string path = TempPath();
    std::remove(path.c_str());
    HighScoreStore store(path);
    store.Update(12);

    std::ifstream in(path);
    std::string contents((std::istreambuf_iterator<char>(in)), std::istreambuf_iterator<char>());
    CHECK(contents.find("12") != std::string::npos);
    std::remove(path.c_str());
}

TEST_CASE("update ignores a lower score") {
    std::string path = TempPath();
    std::remove(path.c_str());
    HighScoreStore store(path);
    store.Update(12);
    store.Update(5);
    CHECK(store.Value() == 12);
    std::remove(path.c_str());
}

TEST_CASE("a new store reloads the persisted value") {
    std::string path = TempPath();
    std::remove(path.c_str());
    HighScoreStore first(path);
    first.Update(9);

    HighScoreStore reloaded(path);
    CHECK(reloaded.Value() == 9);
    std::remove(path.c_str());
}
